### DISCUSSION TITLE
Rendering MDX at runtime using Snowpack fails

### DISCUSSION THREAD

**Discussion by martonlanga** [mdx-js/mdx#1348] [Q&A] [answered]
> Cross-posting from https://github.com/snowpackjs/snowpack/discussions/1674, not sure if this is Snowpack related

Trying to render the `MDX` from `@mdx-js/runtime` results in the following error (_just_ importing it doesn't break it):
```
Uncaught TypeError: (0 , _index.default) is not a function
    at NodePath.traverse (runtime.js:210735)
    at Scope.crawl (runtime.js:193475)
    at Scope.init (runtime.js:193443)
    at NodePath.setScope (runtime.js:209597)
    at NodePath.setContext (runtime.js:209613)
    at new File (runtime.js:227321)
    at normalizeFile (runtime.js:231196)
    at normalizeFile.next (<anonymous>)
    at run (runtime.js:234629)
    at run.next (<anonymous>)
```

[Here's the example from the MDX documentation I'm trying to implement.](https://mdxjs.com/advanced/runtime)

**Note** that there is a snowpack-mdx plugin, but that's for generating components from `.md` & `.mdx` and I'm trying to render `.mdx` during _runtime_.

#### Things I've tried
 - At first I was getting node-polyfill errors, but enabling `polyfillNode` solved those issues
 - Created a blank snowpack-react app, added only `@mdx-js/runtime` and got the same error
 - Did the same using `create-react-app` and it works

[Link to the `@mdx-js/runtime` package](https://github.com/mdx-js/mdx/blob/master/packages/runtime/src/index.js)

**Comment by ChristianMurphy** [comment, marked as answer]
@martonlanga could you share a gist or a github repo with a complete example of the issue you are seeing?

**Reply by martonlanga** [reply]
Sure @ChristianMurphy: https://github.com/martonlanga/snowpack-mdx-runtime

To run:
```
yarn
yarn start
```

`MDX` related code is in `App`. Please let me know if you have any other questions!

Update: just pushed the changes with the MDX runtime example

**Reply by ChristianMurphy** [reply]
Thanks @martonlanga!
Spinning up the repo and looking at the error, it's coming from

*web-modules/@mdx-js/runtime.js*
```js
  buildCodeFrameError(msg, Error = SyntaxError) {
    return this.hub.buildError(this.node, msg, Error);
  }

  traverse(visitor, state) {
>   (0, _index.default)(this.node, visitor, this.scope, state, this);
        ^^^^^^^^^^^^^^
  }

  set(key, node) {
    t.validate(this.node, key, node);
    this.node[key] = node;
  }
```

This piece of code doesn't appear in the distributed package https://unpkg.com/@mdx-js/runtime@1.6.21/dist/cjs.js or https://unpkg.com/@mdx-js/runtime@1.6.21/dist/esm.js it appears it is being injected/bundled by skypack,
It looks like a piece of Babel or Buble. :thinking: 
Which is surprising that it is bundled since:

> Snowpack’s dev server is an instant dev environment for unbundled development.

https://www.snowpack.dev/#snowpack%E2%80%99s-dev-server 
and it looks like it's bundling files :sweat_smile: 

Babel/Buble are used internally by MDX to convert JSX to plain JS, so it makes some sense that it would be a part of the project.
I'm not as sure why skypack is bundling in this case, or why the import isn't being resolved after it bundles.

**Reply by martonlanga** [reply]
Thanks for taking a look @ChristianMurphy!

Hmm, not sure what's going on, maybe some Snowpack guys will know what to do.

Do you have any ideas what I could might try to make this work? Should I try to see whether Babel/Buble work individually?

**Reply by ChristianMurphy** [reply]
A few ideas, if possible considering using https://github.com/jaredLunde/snowpack-plugin-mdx

Runtime can be useful in some scenarios, like live editing, but it comes with the tradeoff:

> ⚠️ warning: this is not the preferred way to use MDX since it introduces a substantial amount of overhead and dramatically increases bundle sizes. It should also not be used with user input that isn’t sandboxed

https://mdxjs.com/advanced/runtime

---

For debugging runtime, I'd start trying lower level utilities, `remark` and `@mdx-js/mdx` in particular, to see if those can run in isolation with snowpack, or if it's something specific to the runtime.
I'm also interested to hear insights from the snowpack team on why this is being bundled, instead of being left unbundled :thinking:

**Reply by martonlanga** [reply]
I'm running cli commands and trying to transform their `stdout` into `mdx`, so it has to be runtime (or?)

---

Will try to make `remark` & `@mdx-js/mdx` work, will let you know how it goes!

**Reply by ChristianMurphy** [reply]
> I'm running cli commands and trying to transform their stdout

Could you expand on this?
When/how are the CLI commands running?
Could their output be written to disk as a file, during/before build time, and transformed by https://github.com/jaredLunde/snowpack-plugin-mdx?

**Reply by martonlanga** [reply]
Sure. I'm creating an Electron app that functions kind of like a terminal and lets users run commands. There are two types:
 1. Regular commands
    Output is plain text (rendered by [xterm.js](https://xtermjs.org/)). (e.g.: `npm install`, `git status`)
 2. _Special_ commands/scripts
    Output is MDX, render at runtime.

Since we don't know anything about the outputs at build time, I think the only option is to transform these at runtime.

**Reply by ChristianMurphy** [reply]
Gotcha, yep, that is what `runtime` is for

**Reply by martonlanga** [reply]
`remark` seems to work, `@mdx-js/mdx` results in the same error:
```js
import React from 'react';
import remark from 'remark';
import html from 'remark-html';
import mdx from '@mdx-js/mdx';

remark()
  .use(html)
  .process('## Hello world!', function (err, file) {
    console.log(String(file));
    // works: <h2>Hello world!</h2>
  });

async function transform() {
  const result = await mdx(`
  # Hello, MDX
  `);

  console.log(result);
  // mdx.js:55587 Uncaught (in promise) TypeError: (0 , _index.default) is not a function
}

export default () => {
  transform();
  return null;
};
```

**Reply by martonlanga** [reply]
Tried `@mdx-js/runtime` with Parcel, getting the following error:
```
index-browser.js:1 Uncaught ReferenceError: regeneratorRuntime is not defined
    at Object.parcelRequire.../node_modules/remark-mdx/node_modules/@babel/core/lib/config/files/index-browser.js (index-browser.js:1)
    at newRequire (src.e31bb0bc.js:47)
    at localRequire (src.e31bb0bc.js:53)
    at Object.parcelRequire.../node_modules/remark-mdx/node_modules/@babel/core/lib/index.js../transformation/file/file (index.js:181)
    at newRequire (src.e31bb0bc.js:47)
    at localRequire (src.e31bb0bc.js:53)
    at Object.parcelRequire.../node_modules/remark-mdx/extract-imports-and-exports.js.@babel/core (extract-imports-and-exports.js:1)
    at newRequire (src.e31bb0bc.js:47)
    at localRequire (src.e31bb0bc.js:53)
    at Object.parcelRequire.../node_modules/remark-mdx/index.js.is-alphabetical (index.js:3)
```

**Reply by martonlanga** [reply]
- Tried porting the `mdx` lib to ES modules and including it in the Snowpack as, but [getting the same `not a function` error](https://github.com/martonlanga/snowpack-mdx-runtime/blob/576d82fab121fedb43b50fd5e49b58050110eef2/src/mdx.js#L111)
- Using Snowpack's [babel plugin didn't help either](https://github.com/martonlanga/snowpack-mdx-runtime/blob/babel/snowpack.config.js)
- Runtime in a Next.js app works

Also thought about running transform in Electron's main process (I have a separate webpack bundler for this part) and sending the result as a string to the fronted. But this seems unnecessarily complex. I would have to include React and my components both in the Electron and Snowpack part of my app.

Then I came across `markdown-to-jsx`, which seems to work. Read your [comparison of the two](https://spectrum.chat/mdx/general/what-are-the-advantages-of-using-markdown-to-jsx-over-remark-mdx-or-mdx~133cb3ed-ded7-42c7-bf4d-dff328e22df7?m=MTU5NDY1ODE2Mzc0Ng==) and missing out on transformers is fine for now I guess. Also saw you mentioned the work being done on `micromark`. Do you think `mdx` using `micromark` might fix this issue in the future?

Thanks for your time @ChristianMurphy - and sorry for bugging you so much!

**Reply by ChristianMurphy** [reply]
> ```
> index-browser.js:1 Uncaught ReferenceError: regeneratorRuntime is not defined
>    at Object.parcelRequire.../node_modules/remark-mdx/node_modules/@babel/core/lib/config/files/index-browser.js (index-browser.js:1)
>    at newRequire (src.e31bb0bc.js:47)
>    at localRequire (src.e31bb0bc.js:53)
>    at Object.parcelRequire.../node_modules/remark-mdx/node_modules/@babel/core/lib/index.js../transformation/file/file (index.js:181)
>    at newRequire (src.e31bb0bc.js:47)
>    at localRequire (src.e31bb0bc.js:53)
>    at Object.parcelRequire.../node_modules/remark-mdx/extract-imports-and-exports.js.@babel/core (extract-imports-and-exports.js:1)
>    at newRequire (src.e31bb0bc.js:47)
>    at localRequire (src.e31bb0bc.js:53)
>    at Object.parcelRequire.../node_modules/remark-mdx/index.js.is-alphabetical (index.js:3)
> ```

https://unpkg.com/remark-mdx@1.6.21/index.js doesn't include regenerator runtime,
it sounds like https://babeljs.io/docs/en/babel-runtime is coming into the mix somehow :thinking: 

> * Tried porting the mdx lib to ES modules and including it in the Snowpack as, but getting the same not a function error
> * Using Snowpack's babel plugin didn't help either
> * Runtime in a Next.js app works

:thinking: strange, that make it sound like something may be interfering with snowpack's package resolver directly :thinking: 

> Do you think mdx using micromark might fix this issue in the future?

It may, I know @wooorm as been considering the possibility of removing babel as a dependency/peer-dependency of the project.
There's also the possibility it's an issue in snowpack's package resolver, which may get resolved there, and unblock usage here.

**Reply by wooorm** [reply]
This much more sounds like a crash with the things snowpack is doing, rather than what we’re doing. It also might be a “working as intended” for them: `eval` is dangerous.

I don’t think it has to do with how MDX is parsed, so it’s unrelated to micromark/remark. But, for micromark: import/export/jsx is still supported in the MDX format (for the JSX specifically I mean: `# Hello {<Icon />}`), so you’ll have to transform it with something.
What I plan on working on is outputting JS instead of JSX from `@mdx-js/mdx`. Which might save one compile.

Comparing `markdown-to-jsx` to MDX is somewhat like apples and oranges. Why do you want to use JSX / MD / MDX?

**Reply by martonlanga** [reply]
> There's also the possibility it's an issue in snowpack's package resolver, which may get resolved there, and unblock usage here.
> This much more sounds like a crash with the things snowpack is doing, rather than what we’re doing. It also might be a “working as intended” for them: eval is dangerous.

I'll try to reach out to someone at Snowpack to find out whether this not working, is a bug or a feature. Will update!

> What I plan on working on is outputting JS instead of JSX from @mdx-js/mdx. Which might save one compile.

That sounds interesting. I'm watching the releases so gonna give that a try.

> Why do you want to use JSX / MD / MDX?

@wooorm I'm creating an Electron app that lets users run commands and converts `stdout` to `JSX`. So essentially want to give CLIs the ability to output MDX. I know there's https://mdxjs.com/guides/terminal, which is pretty cool but what if I want to display cat gifs haha
